src/iter/collect/consumer.rs

```diff
@@ -4,16 +4,34 @@ use std::mem::MaybeUninit;
 use std::ptr;
 use std::slice;
 
+/// We need to store raw pointers in a `Send` struct to remember
+/// provenance (see `CollectResult`).
+#[derive(Clone, Copy)]
+struct SendPtr<T>(*mut T);
+
+// SAFETY: !Send for raw pointers is not for safety, just as a lint
```

Comment:
Is this really correct? Raw pointers could point into thread-local data or short-lived stack data, couldn't they?

Reply:
They sure can. But you can make any raw pointer Send by wrapping it in AtomicPtr using 100%-safe code. The safety comment I wrote is based on: https://doc.rust-lang.org/nomicon/send-and-sync.html?highlight=Lint#send-and-sync

Reply:
I think that's ok, but we probably want `T: Send` here for it to be safe to dereference on another thread. The rest of the code already has that, but I think it would be cleaner to have it here too.

It could also use `NonNull` to get the niche, but we don't want covariance, so it would take another `PhantomData` to counteract that.